### ISSUE TITLE
Persist Charge (2/4) - Stop flow

### ISSUE THREAD

**Issue by prtandrup** [Energinet-DataHub/geh-charges#1153]
**Synopsis:**
As a Grid Access Provider,
I want to stop a Charge
So that I keep the master data aligned with my own system

This is another story of a few stories with the goal of implementing the [charge persistence rule set](https://github.com/Energinet-DataHub/geh-charges/tree/main/docs/process-flows#persist-charge) (master data only)

With this story we will implement the **stop** flow, highlighted below:
<hr>

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/2439e89c-49ce-434c-a497-4e4b00dc1917)

Notes:
In a stop request; effective date + termination date are set and equal.

**Acceptance criteria:**

- [x] 1 Given a charge exists with multiple charge periods
When a stop request is received with a stop date inside the first charge period
Then the charge is stopped on stop date
And the persisted result must be like:

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/f23c0225-0199-48ac-a26b-39508cb407c5)

- [x] 2 Given a charge exists with multiple charge periods
When a stop request is received with a stop date equal to the start date of the third period
Then the charge is stopped on the stop date 
Ad the persisted result must be like:

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/a8c9ae1e-7d63-4618-8ae5-68b8ffa91f9a)

- [x] 3 Given a charge exists with two periods and the charge is stopped
When a stop request is received with a stop date inside the first charge period
Then charge is stopped on the new stop date
And the persisted result must be like:

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/f5d4eb0f-4a09-4903-a2f5-63cbebb5a795)

- [x]  4 Given a charge exists with prices
When a stop request is received
Then the charge is stopped
And persisted prices from the stop date and onwards are removed

**Comment by prtandrup** [comment]
### Testing AC# 1 in U-001.

- [ ] 1 Given a charge exists with multiple charge periods
When a stop request is received with a stop date inside the first charge period
Then the charge is stopped on stop date
And the persisted result must be like:

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/f23c0225-0199-48ac-a26b-39508cb407c5)

**Starting point:**
![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/96434999-cfd0-432e-9fd4-d99f11fb61b4)

**Submitted stop @ 2022-05-03T22:00:00**

````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
	<cim:mRID>DocId2022-03-18T13:23:47.802Z</cim:mRID>
	<cim:type>D10</cim:type>
	<cim:process.processType>D18</cim:process.processType>
	<cim:businessSector.type>23</cim:businessSector.type>
	<cim:sender_MarketParticipant.mRID codingScheme="A10">9656626091925</cim:sender_MarketParticipant.mRID>
	<cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
	<cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
	<cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
	<cim:createdDateTime>2022-03-18T14:23:47Z</cim:createdDateTime>
	<cim:MktActivityRecord>
		<cim:mRID>OpId2022-03-18T13:23:47.802Z</cim:mRID>
		<cim:ChargeGroup>
			<cim:ChargeType>
				<cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">9656626091925</cim:chargeTypeOwner_MarketParticipant.mRID>
				<cim:type>D03</cim:type>
				<cim:mRID>ChgId009</cim:mRID>
				<cim:name>Charge Tariff day Name</cim:name>
				<cim:description>The charge description update</cim:description>
				<cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
				<cim:effectiveDate>2022-05-03T22:00:00Z</cim:effectiveDate>
                <cim:terminationDate>2022-05-03T22:00:00Z</cim:terminationDate>
				<cim:VATPayer>D02</cim:VATPayer>
				<cim:transparentInvoicing>false</cim:transparentInvoicing>
				<cim:taxIndicator>true</cim:taxIndicator>
			</cim:ChargeType>
		</cim:ChargeGroup>
	</cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

**Result:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/cb5320dc-551e-4031-bba2-aa1dda3922f4)

Notification also looks OK:

````
<?xml version="1.0" encoding="utf-8"?>
<cim:NotifyPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:notifypricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:notifypricelist:0:1 urn-ediel-org-structure-notifypricelist-0-1.xsd">
    <cim:mRID>b05d118cf7634548b8f9681cebef8d9a</cim:mRID>
    <cim:type>D12</cim:type>
    <cim:process.processType>D18</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDZ</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">9656626091925</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDM</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>2022-03-18T13:25:03Z</cim:createdDateTime>
    <cim:MktActivityRecord>
        <cim:mRID>38a9c3455b124e819ef9a432682ea7b6</cim:mRID>
        <cim:snapshot_DateAndOrTime.dateTime>2022-03-18T13:23:45Z</cim:snapshot_DateAndOrTime.dateTime>
        <cim:ChargeGroup>
            <cim:ChargeType>
                <cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">9656626091925</cim:chargeTypeOwner_MarketParticipant.mRID>
                <cim:type>D03</cim:type>
                <cim:mRID>ChgId009</cim:mRID>
                <cim:name>Charge Tariff day Name</cim:name>
                <cim:description>The charge description update</cim:description>
                <cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
                <cim:effectiveDate>2022-05-03T22:00:00Z</cim:effectiveDate>
                <cim:terminationDate>2022-05-03T22:00:00Z</cim:terminationDate>
                <cim:VATPayer>D02</cim:VATPayer>
                <cim:transparentInvoicing>false</cim:transparentInvoicing>
                <cim:taxIndicator>true</cim:taxIndicator>
            </cim:ChargeType>
        </cim:ChargeGroup>
    </cim:MktActivityRecord>
</cim:NotifyPriceList_MarketDocument>
````

**Comment by prtandrup** [comment]
### Testing AC# 2 in U-001.

- [ ] 2 Given a charge exists with multiple charge periods
When a stop request is received with a stop date equal to the start date of the third period
Then the charge is stopped on the stop date 
Ad the persisted result must be like:

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/a8c9ae1e-7d63-4618-8ae5-68b8ffa91f9a)

**Staring point:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/5f4e9805-ef02-4e5e-bd28-4f9c6d494bab)

**Submitted stop @ 2022-05-10T22:00:00**

````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
	<cim:mRID>DocId2022-03-18T13:47:06.365Z</cim:mRID>
	<cim:type>D10</cim:type>
	<cim:process.processType>D18</cim:process.processType>
	<cim:businessSector.type>23</cim:businessSector.type>
	<cim:sender_MarketParticipant.mRID codingScheme="A10">9656626091925</cim:sender_MarketParticipant.mRID>
	<cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
	<cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
	<cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
	<cim:createdDateTime>2022-03-18T14:47:06Z</cim:createdDateTime>
	<cim:MktActivityRecord>
		<cim:mRID>OpId2022-03-18T13:47:06.365Z</cim:mRID>
		<cim:ChargeGroup>
			<cim:ChargeType>
				<cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">9656626091925</cim:chargeTypeOwner_MarketParticipant.mRID>
				<cim:type>D03</cim:type>
				<cim:mRID>ChgId010</cim:mRID>
				<cim:name>Charge Tariff day Name</cim:name>
				<cim:description>The charge description update</cim:description>
				<cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
				<cim:effectiveDate>2022-05-10T22:00:00Z</cim:effectiveDate>
                <cim:terminationDate>2022-05-10T22:00:00Z</cim:terminationDate>
				<cim:VATPayer>D02</cim:VATPayer>
				<cim:transparentInvoicing>false</cim:transparentInvoicing>
				<cim:taxIndicator>true</cim:taxIndicator>
			</cim:ChargeType>
		</cim:ChargeGroup>
	</cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

**Result:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/c63ac327-641d-4dc0-a333-dbd4110d13ad)

**Comment by prtandrup** [comment]
@jonasdmoeller, what happened to the boolean "IsStopped" Charge variable we discussed a bit back? Did you decide not to go that way?

**Comment by prtandrup** [comment]
notes from team talk this morning:
- there is no bool 'IsStopped' in the storage layer.
- when stopping a charge, the latest period's end date will be set to the requested stop date (no row with equal start and end dates)
- the Charge domain model will be able to answer "IsStopped"

**Comment by prtandrup** [comment]
### Testing AC# 3 in U-001:

- [ ] 3 Given a charge exists with two periods and the charge is stopped
When a stop request is received with a stop date inside the first charge period
Then charge is stopped on the new stop date
And the persisted result must be like:

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/f5d4eb0f-4a09-4903-a2f5-63cbebb5a795)

**Starting point:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/629774bb-6e6b-4ab6-bf8c-d0a61f1cc33f)

**Submitted new stop date = 2022-05-04T22:00:00Z**
````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
    <cim:mRID>DocId{{$isoTimestamp}}</cim:mRID>
    <cim:type>D10</cim:type>
    <cim:process.processType>D18</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">{{GridAccessProvider}}</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>{{ISO8601Timestamp}}</cim:createdDateTime>
    <cim:MktActivityRecord>
        <cim:mRID>OpId{{$isoTimestamp}}</cim:mRID>
        <cim:ChargeGroup>
            <cim:ChargeType>
                <cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">{{GridAccessProvider}}</cim:chargeTypeOwner_MarketParticipant.mRID>
                <cim:type>D03</cim:type>
                <cim:mRID>ChgId012</cim:mRID>
                <cim:name>Charge Tariff day Name</cim:name>
                <cim:description>The charge description update</cim:description>
                <cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
                <cim:effectiveDate>2022-05-04T22:00:00Z</cim:effectiveDate>
                <cim:terminationDate>2022-05-04T22:00:00Z</cim:terminationDate>
                <cim:VATPayer>D02</cim:VATPayer>
                <cim:transparentInvoicing>false</cim:transparentInvoicing>
                <cim:taxIndicator>true</cim:taxIndicator>
            </cim:ChargeType>
        </cim:ChargeGroup>
    </cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

**Result:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/63dca398-a962-4063-ad6a-bbeb23064878)

**Comment by prtandrup** [comment]
### Testing AC# 4 in U-001 

- [ ]  4 Given a charge exists with prices
When a stop request is received
Then the charge is stopped
And persisted prices from the stop date and onwards are removed

OBS: This AC can only be partly verified due to the following 2 reasons:
* Prices (aka point) are only store on 'Create' actions, as persisting prices rule set is yet to be defined and implemented (story #1196)
* VR.507-1 currently blocks submissions of price series longer than a single day (story #1198)

**Starting point:**
Hence, the starting point will be a bit alternative.
1. Charge created at 2022-05-05T22:00:00 with a price
2. Charge updated at 2022-05-01T22:00:00 (no price stored)

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/7099b37c-9734-4413-92ed-050ee23ddc5a)

**Stopping charge _on_ Price date (2022-05-05T22:00:00) **

````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
    <cim:mRID>DocId{{$isoTimestamp}}</cim:mRID>
    <cim:type>D10</cim:type>
    <cim:process.processType>D18</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">{{GridAccessProvider}}</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>{{ISO8601Timestamp}}</cim:createdDateTime>
    <cim:MktActivityRecord>
        <cim:mRID>OpId{{$isoTimestamp}}</cim:mRID>
        <cim:ChargeGroup>
            <cim:ChargeType>
                <cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">{{GridAccessProvider}}</cim:chargeTypeOwner_MarketParticipant.mRID>
                <cim:type>D03</cim:type>
                <cim:mRID>ChgId014</cim:mRID>
                <cim:name>Charge Tariff day Name</cim:name>
                <cim:description>The charge description update</cim:description>
                <cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
                <cim:effectiveDate>2022-05-05T22:00:00Z</cim:effectiveDate>
                <cim:terminationDate>2022-05-05T22:00:00Z</cim:terminationDate>
                <cim:VATPayer>D02</cim:VATPayer>
                <cim:transparentInvoicing>false</cim:transparentInvoicing>
                <cim:taxIndicator>true</cim:taxIndicator>
            </cim:ChargeType>
        </cim:ChargeGroup>
    </cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

**Result:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/08a5755b-e4ab-4934-8c88-c4d697b39773)

#### Additional test - stop _before_ price date

Starting point: 

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/63e13a1c-5d64-48ac-a006-ebcd76310c77)

**Stopping charge _before_ Price date (2022-05-04T22:00:00) **

````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
    <cim:mRID>DocId{{$isoTimestamp}}</cim:mRID>
    <cim:type>D10</cim:type>
    <cim:process.processType>D18</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">{{GridAccessProvider}}</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>{{ISO8601Timestamp}}</cim:createdDateTime>
    <cim:MktActivityRecord>
        <cim:mRID>OpId{{$isoTimestamp}}</cim:mRID>
        <cim:ChargeGroup>
            <cim:ChargeType>
                <cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">{{GridAccessProvider}}</cim:chargeTypeOwner_MarketParticipant.mRID>
                <cim:type>D03</cim:type>
                <cim:mRID>ChgId015</cim:mRID>
                <cim:name>Charge Tariff day Name</cim:name>
                <cim:description>The charge description update</cim:description>
                <cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
                <cim:effectiveDate>2022-05-04T22:00:00Z</cim:effectiveDate>
                <cim:terminationDate>2022-05-04T22:00:00Z</cim:terminationDate>
                <cim:VATPayer>D02</cim:VATPayer>
                <cim:transparentInvoicing>false</cim:transparentInvoicing>
                <cim:taxIndicator>true</cim:taxIndicator>
            </cim:ChargeType>
        </cim:ChargeGroup>
    </cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

**Result:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/20d0b335-511b-47cb-87c1-a428fbd09a9c)

#### Additional test - stop _after_ price date

**Starting point:**

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/3eaa7ece-d7cd-49f6-842b-4097ee51b050)

**Stopping charge _before_ Price date (2022-05-07T22:00:00) **

![image.png](https://images.zenhubusercontent.com/60ae42d0916bb92d8b000901/86cd554b-b919-4550-b5ee-1c3ee7917d62)
